rfc-134-govuk-wide-session-cookie-and-login.md

```diff
@@ -0,0 +1,354 @@
+
+### Non-personalised parts of GOV.UK won't manipulate the session cookie
+
+It's unlikely that all of GOV.UK will be personalised, so there will
+be islands of personalised content.  Currently there is the Transition
+Checker.  Maybe next will be some guidance pages, or something else.
+We want to keep the user's session alive while they are browsing the
+non-personalised parts of GOV.UK, otherwise we risk this bad
+experience:
+
+1. The user logs in to use some personalised part of GOV.UK.
+
+2. The user then spends 30 minutes (or whatever we use for the session
+   duration) viewing non-personalised parts of GOV.UK, but without
+   leaving the site.
+
+3. The user then tries to use another personalised part of GOV.UK, but
+   their session has expired, because the non-personalised parts
+   weren't bumping the expiration time on every page view.
+
+### We can't cache as effectively
+
+The Fastly docs have some [comments on the risks of cookies][fastly-cookies-risks]:
+
+> Cookies can lead to undesirable outcomes.  At worst, if a cookie
+> header is forwarded to your backend server, the backend uses the
+> cookie value to generate personalized content, and that content is
+> then cached by Fastly, a user may end up receiving content intended
+> for someone else.  A theoretical solution to this, adding a Vary:
+> Cookie header to the response, leads to another bad outcome: the
+> response is most likely not cacheable at all, and Fastly will
+> forward all requests to your backend.
+
+[fastly-cookies-risks]: https://developer.fastly.com/reference/http-headers/Cookie/
+
+## Proposal
+
+### Set a session cookie on www.gov.uk
+
+We'll set a new cookie, `__Host-govuk_account_session`, on www.gov.uk.
+This cookie will hold a unique session ID, randomly generated (in a
+collision-resistant way) when the user signs in, which is encrypted
+using asymmetric crypto.
+
+This cookie is a secure, [domain-locked][], session cookie:
+
+```
+Set-Cookie: __Host-govuk_account_session=<value>; secure; httponly; samesite=strict; path=/
+```
+
+It can be expired, logging a user out, by re-setting it with
+`max-age=0`.
+
+This cookie cannot be set on the `gov.uk` domain, so service domains
+will need to re-authenticate and manage their own sessions.  This RFC
+is just about authentication and attribute use on www.gov.uk itself.
+The cross-government single-sign-on part of this work is out of scope.
+
+[domain-locked]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#Cookie_prefixes
+
+#### Cookie encryption
+
+The cookie value is encrypted with a private key.  This ensures that
+an attacker cannot tamper with the value.  Apps which consume the
+cookie can use the public key to validate it.  This key can be truly
+public, as the session ID itself is not sensitive.
+
+If the cookie value is malformed, for example if an attacker has tried
+re-encrypting it with a different private key, the consuming app
+should log the user out by expiring the cookie.
```

Comment:
If the value of the cookie is a securely generated random number with sufficient entropy, I'm not sure we need any of this cryptography. It should already be tamper proof, by virtue of the attacker not being able to guess another random number.

If we do need crypto, I think we want signing here, not encryption. "This ensures that an attacker cannot tamper with the value" is not true for many encryption schemes (for example AES-CBC is [malleable](https://en.wikipedia.org/wiki/Malleability_%28cryptography%29), AES-GCM doesn't have this problem, but it's still the wrong tool for the job).

You could still use an asymmetric signing algorithm (like ECDSA), but I'm really not sure what that buys us if we assume the attacker is somehow capable of guessing any random numbers we generate.

Reply:
@OllieJC  - just to check I'm not misunderstanding something?

Reply:
The purpose of crypto is to assert that sign-in created the cookie, rather than another process iterating through session IDs and trying to see if they're a valid session.

Signatures are fine as long as they're always checked - an upshot of encryption is that it enforces the use of a public key to get the session ID so no origin can use the session ID without decrypting (vs. the signature being potentially optional)

Reply:
```suggestion
The cookie value is encrypted with a private key. This ensures that
an attacker cannot re-encrypt the value without knowing the private
key - which is never shared. Apps which consume the cookie can 
use the public key to decrypt, essentially validating the source.
This public key can be truly public, as the session ID itself is not
sensitive, but we do need some way of validating it came from the
right place.

If the cookie value is malformed, for example if an attacker has tried
re-encrypting it with a different private key, the consuming app
shouldn't be able to decrypt it with the public key and should log
the user out by expiring the cookie.
```
Or:
```suggestion
The cookie value is the session ID and a signature of that ID signed
by a private key. This ensures that an attacker cannot re-sign the ID
without knowing the private key - which is never shared with apps.
Apps which consume the cookie must check the signature with the
public key, validating the source. This public key can be truly public,
as the session ID itself is not sensitive, but we do need some way of
validating it came from the right place.

If the cookie value is malformed, for example if an attacker has tried
re-signing it with a different private key or not including a signature,
the consuming app must ignore it and should log the user out by
expiring the cookie.
```

Reply:
I'm still not sure it's worth the extra complexity.

Let's say we generate our session IDs with `SecureRandom.hex(128/8)` (a random 128 bit number).

If we've got 2^32 active sessions (that's ~4 billion), and we let an attacker make 2^32 guesses per second (a few guesses per nanosecond), it's still going to take them in the order of 2^64 seconds (585 billion years) to guess a valid session ID. Want more security? Use a 256 bit random number.

I suppose being able to validate the cookie before hitting the database to check it's a real session might be nice, but I would have thought a (symmetric) HMAC would be more appropriate for that. Yes, we'd have to share the HMAC keys between the thing doing the signing and the things doing the validation. But even if an HMAC secret got leaked, it would only let an attacker guess random numbers, which isn't going to help them.

Or perhaps you're thinking about a situation where an attacker has somehow managed to get read access to the database where the session IDs are stored, but doesn't have access to the private key that would let them forge new cookies because that's stored somewhere more secure? To be honest, I haven't really considered this from a "some of the infrastructure is compromised and it still needs to be secure" perspective.

Reply:
Yeah that's totally reasonable. I don't even think you need a 128bits of entropy and it still be implausible to take a wild guess.

My primary thinking around using a signature is that it asserts to a consuming app verifying it that it came from the signing in mechanism. If in future an ID is leaked in anyway, or can be fetched out using an API etc. then it cannot be used by to gain access, as that requires the secret. As for asymmetric, this means the public key can be truly public and not kept on an app. Knowledge of the public key doesn't allow session cookies can be forged.

For example, have an endpoint like:
`/.well-known/session-public-key-158.pub`
If the value of the cookie was something like:
`158:987654321:hash`
Then the consumer app can fetch the public key, cache it, hash the ID and compare the hashes.

I think there are some nuanced attacks here, it'll be best to threat model the proposal before any implementation - I'll put a Jamboard together tomorrow to try and surface some possible attacks together asynchronously.

Reply:
A simple way of generating a strong random collision-resistant session ID would be something like this:

```ruby
def generate_session_id
  "#{Time.zone.now.to_f}-#{SecureRandom.hex(128/8)}"
end
```

- `Time.zone.now.to_f` gives the current Unix time in microseconds
- `SecureRandom.hex(128/8)` gives a 128-bit integer in hexadecimal

So we'd end up with session IDs like `1612452968.1587293-b74b7fd4077caf6cabf6632736313653`

Given that the cookie is using the `__Host-` prefix, it seems unlikely that an attacker would be able to obtain the user's session ID without directly nabbing the user's cookie (eg, through malware on their computer, or a MitM attack), and if they can do that then an encrypted or signed cookie doesn't help us.

So, so long as we don't expose the session ID to javascript (eg, including it in the page body), does encrypting it gain us much?

Reply:
I'm going to resolve this because we've had a talk about it and changed the RFC.  If the new way is still bad we can start another thread.